OpenSim/Simulation/GeodesicWrap/GeodesicWrapObject.h

```diff
@@ -0,0 +1,84 @@
+#ifndef GEODESIC_WRAP_OBJECT_H_
+#define GEODESIC_WRAP_OBJECT_H_
+
+#include "GeodesicCurveState.h"
+#include "ImplicitSurfaceParams.h"
+#include <memory>
+#include <vector>
+
+namespace OpenSim
+{
+
+    // TODO rename WrapObject to Surface?
+
+//==============================================================================
+// GEODESIC WRAP OBJECT INTERFACE
+//==============================================================================
+/// This is where implicit and parametric surfaces come together.
+// Also provides the separation between local and global coordinates.
+class GeodesicWrapObjectImpl
+{
+public:
+    virtual void calcInitState(
+        const SimTK::Vec3& pStart,
+        const SimTK::Vec3& pEnd) = 0;
+
+    // Shoots the geodesic over the surface.
+    virtual void calcLocalGeodesic(
+        GeodesicCurveState& start,
+        GeodesicCurveState& end,
+        double& length,
+        std::vector<SimTK::Vec3>& pointsLog) = 0;
+
+    virtual void applyVariation(
+        const GeodesicVariation& var) = 0;
+};
+
+//==============================================================================
+// CONCRETE GEODESIC WRAP OBJECT
+//==============================================================================
+// Handles both implicit, parametric and analytic.
+// Applies local-surface-coords to global-coords
+// Does the shooting: calcGeodesic.
+class GeodesicWrapObject
+{
+    public:
+    // Construct Implicit GeodesicWrapObject
+    GeodesicWrapObject(
+            ImplicitSurfaceParams&& surface,
+            std::shared_ptr<SimTK::Transform> transformHandle
+            );
+
+    // Construct Parametric GeodesicWrapObject
+    /* GeodesicWrapObject( */
+    /*         ParametricSurfaceParams&& surface, */ 
+    /*         std::shared_ptr<SimTK::Transform> transformHandle */
+    /*         ); */
+
+
+    private:
+    GeodesicWrapObject(
+            std::unique_ptr<GeodesicWrapObjectImpl>&& impl,
+            std::shared_ptr<SimTK::Transform> transform
+            );
+
+    public:
+
+    // Shoots the geodesic.
+    // Applies conversion from local to global coordinates.
+    void calcGeodesic(GeodesicCurve& shooterResult);
+
+    void applyVariation(
+        const GeodesicVariation& var);
+
+    // Triggers fetching the global coordinates from the underlying surface body.
+    const SimTK::Transform& updTransformToGround();
+
+private:
+    std::unique_ptr<GeodesicWrapObjectImpl> _impl;
+    std::shared_ptr<SimTK::Transform> _transform; // TODO a socket or something?
```

Comment:
If `GeodesicWrapObject` isn't a `ModelComponent` (which it probably will not be), then we can't use a `Socket`. But we could create a `SimTK::ReferencePtr` to the `Frame` based on the `Socket` in the `Surface` class. We could assign the ref pointer when constructing the `GeodesicWrapObject` given a `Surface` object. 

This would mean we would need `GeodesicPath` to hang on to the `Surface`s it needs to construct the `GeodesicWrapObjects`. _That_ we could do using a list `Socket`, since `GeodesicPath` is a `ModelComponent`.

Reply:
I understood that we can also give the `GeodesicWrapObject` the index of the body that is its base. The mobilizer body index? I still have to check it out, I think I've seen you use it.

Would this simplify linking it to the correct body? Or does it still have to be a `Component` then?

Reply:
Ah yes, a `MobilizedBodyIndex` would be better! The transform can then be calculated via [MobilizedBody::getBodyTransform()](https://simbody.github.io/3.7.0/classSimTK_1_1MobilizedBody.html#af05e28bfed85a73e12b04f8d8f2a06a0) when needed.